### ISSUE TITLE
Redacted Url - Remove Personally-Identifying Information

### ISSUE THREAD

**Issue by jackiemunroe** [lemonlabsuk/scala-uri#277]
**Description**
There is already the ability to remove query parameters, which is great. But, I think that it would be really useful to also strip any user/password info from the Url for the purpose of removing any personally identifying information. My use case is for report generation and logging.
I think for most that would want to use this feature, it would just return a string with all UserInfo and Query Params removed. Not necessarily required though. I have a couple of ideas of how this would be used, but as a scala novice, I am unsure what the best implementation would be.

My preference would be for Idea 4B.

**Example**
```scala
var redactMe = Url.parse(toBeRedacted)

// idea #1
// leave to user to remove what is needed
redactMe.removeQueryString()
redactMe.removeUser() // set user to Null, if set
redactMe.removePassword() // set password to Null, if set 

logger.info(redactMe.toString())

// idea #2 
// leave to user to remove what is needed, but use single method to remove user + password
redactMe.removeQueryString()
redactMe.removeUserInfo() // set User to Null, if set and set Password to Null if set

logger.info(redactMe.toString())


// idea #3
// Remove anything considered to be Personally-Identifying Information (PII) in one go
redactMe.redactPII() // set query, user, and password to null if set

logger.info(redactMe.toString())


// idea #4 
// Create a method specifically with safe logging in mind, that returns a string with PII removed
// perhaps have options available that all default to true

logger.info(redactMe.toRedactedString()) // query, user and password would all be removed by default

// 4A Option ideas #1
// removeUser - default to true: removes user
// removePassword - default to true: removes password
// removeQueryParams - defaults to true: removes query params
var redactedString = redactMe.toRedactedString(removeUser=True, removePassword=True, removeQueryParams=True)

logger.info(redactedString)


// 4B Option ideas #2
// removeUserInfo - default to true: removes user + password
// removeQueryParams - defaults to true: removes query params
var redactedString = redactMe.toRedactedString(removeUserInfo=True, removeQueryParams=True)

logger.info(redactedString)
```

**Comment by theon** [comment]
Hi @jackiemunroe! thanks for raising this

I agree that a `toRedactedString` would be useful and am happy for this to be added. Two more requirements I think would be useful:
- The ability for the user to specify which parameters (by name) should be redacted
- An optional placeholder which would replace redacted url parts as an alternative to removing them

I was thinking something like below. It is more long-winded than your 4A/4B options, but has the added flexibility for these extra requirements

```scala
val redactorOne = Redact.byRemoving.allParams().user().password()
val redactorTwo = Redact.withPlaceholder("xxx").params("secret", "name").user().password()

val url = Url.parse("http://user:password@example.com?secret=123&other=true")

url.toString(redactorOne) // Returns http://example.com
url.toString(redactorTwo) // Returns http://xxx:xxx@example.com?secret=xxx&other=true
```

It could also be inlined to be a bit closer to your 4A:
```scala
url.toString(Redact.byRemoving.allParams().user().password())
```

Let me know what you think or if you have any other ideas

Side note: I'm surprised that the `Url` type is missing methods to remove/map/replace the user and password, and even the `AbsoluteUrl` type only has the methods to replace the user and password. I'll look to fix this at the same time.

**Comment by jackiemunroe** [comment]
I think that sounds excellent. I greatly appreciate the responsiveness.

**Comment by theon** [comment]
That's released in version `3.1.0` and should be in maven central within a couple hours. I added it to the [README](https://github.com/lemonlabsuk/scala-uri#redacting-urls)

I made one slight tweak from the above. Because a URL with a password, but without a user is invalid, I changed `Redact.byRemoving` so it doesn't allow removing the user without also removing the password. `Redact.byRemoving.userInfo()` removes both user+password and `Redact.byRemoving.password()` just removes the password

Redacted URLs being valid is possibly not very important, but it was easier to implement that way 😄
